tests/unit/task_system_test.c

```diff
@@ -0,0 +1,349 @@
+
+   void *thread_rc;
+   rc = pthread_join(new_thread, &thread_rc);
+   ASSERT_EQUAL(0, rc);
+
+   // This main-thread's thread-index remains unchanged.
+   threadid get_tid_after_thread_exits = platform_get_tid();
+   ASSERT_EQUAL(main_thread_idx,
+                get_tid_after_thread_exits,
+                "main_thread_idx=%lu != get_tid_after_thread_exits=%lu",
+                main_thread_idx,
+                get_tid_after_thread_exits);
+}
+
+/*
+ * Test creation and execution of multiple threads which will do the stuff
+ * required to start using Splinter interfaces (in a real application
+ * code-flow).
+ */
+CTEST2(task_system, test_multiple_threads)
+{
+   pthread_t      new_thread;
+   thread_config  thread_cfg[MAX_THREADS];
+   thread_config *thread_cfgp = NULL;
+   int            tctr        = 0;
+   int            rc          = 0;
+
+   ZERO_ARRAY(thread_cfg);
+   platform_default_log(" Before threads start, task_get_max_tid() = %lu\n",
+                        task_get_max_tid(data->tasks));
+
+   // Start-up n-threads, record their expected thread-IDs, which will be
+   // validated by the thread's execution function below.
+   for (tctr = 1, thread_cfgp = &thread_cfg[tctr];
+        tctr < ARRAY_SIZE(thread_cfg);
+        tctr++, thread_cfgp++)
+   {
+      // These are independent of the new thread's creation.
+      thread_cfgp->tasks          = data->tasks;
+      thread_cfgp->exp_thread_idx = tctr;
+
+      rc =
+         pthread_create(&new_thread, NULL, exec_one_of_n_threads, thread_cfgp);
+      ASSERT_EQUAL(0, rc);
+
+      thread_cfgp->this_thread_id = new_thread;
+   }
+
+   // Complete execution of n-threads. Worker fn does the validation.
+   for (tctr = 1, thread_cfgp = &thread_cfg[tctr];
+        tctr < ARRAY_SIZE(thread_cfg);
+        tctr++, thread_cfgp++)
+   {
+      void *thread_rc;
+      rc = pthread_join(thread_cfgp->this_thread_id, &thread_rc);
+      ASSERT_EQUAL(0, rc);
+   }
+}
+
+/*
+ * Worker routine executed by a single thread. In this method, we also do
+ * careful examination and validation of internal structures to follow the
+ * machinations of the task system.
+ */
+static void *
+exec_one_thread(void *arg)
+{
+   thread_config *thread_cfg = (thread_config *)arg;
+
+   uint64 task_bitmask_before_register =
+      task_active_tasks_mask(thread_cfg->tasks);
+
+   // Verify that the state of active-threads bitmask (managed by Splinter) has
+   // not changed just by creating this pthread. It should be the same as what
+   // we had recorded just prior to pthread_create().
+   ASSERT_EQUAL(thread_cfg->active_threads_bitmask,
+                task_bitmask_before_register);
+
+   // This is the important call to initialize thread-specific stuff in
+   // Splinter's task-system, which sets up the thread-id (index) and records
+   // this thread as active with the task system.
+   task_register_this_thread(thread_cfg->tasks, trunk_get_scratch_size());
```

Comment:
Testing this is fine, but if we use the task thread creation machinery as I suggest above, then this call will happen as a hook.

This function is generally used only for the `main` thread or for threads otherwise created by a non-splinter context. So perhaps we should run this test on the main thread instead.

Reply:
@ajhconway -- Not sure I understand what to change here.

With the revised set of tests, let me know if your comment above is still relevant and do I need to readjust the test code in any way

Reply:
Good now.